File: occasum-core/src/main/java/pw/stamina/occasum/properties/collection/CollectionProperty.java
package pw.stamina.occasum.properties.collection;

import pw.stamina.occasum.properties.Property;

import java.util.Collection;

public interface CollectionProperty<T> extends Property {

    //TODO: String based add and remove methods

    boolean add(T element);

    boolean remove(T element);

    boolean contains(T element);

    Collection<T> getElements();

    static <T> CollectionPropertyBuilder<T> builder(String name) {
        return new CollectionPropertyBuilder<>(name);
    }
}

File: occasum-core/src/main/kotlin/pw/stamina/occasum/properties/collection/CollectionProperty.kt
package pw.stamina.occasum.properties.collection

import pw.stamina.occasum.properties.Property

interface CollectionProperty<T> : Property {

    val elements: Collection<T>

    //TODO: String based add and remove methods

    fun add(element: T): Boolean

    fun remove(element: T): Boolean

    operator fun contains(element: T): Boolean

    companion object {

        fun <T> builder(name: String): CollectionPropertyBuilder<T> {
            return CollectionPropertyBuilder(name)
        }
    }
}
